accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,447 @@
+Given that .NET Tools, when installed on Linux and Mac, have a native shim with an executable bit set, they should be considered valid plugins if the naming convention is followed.
+
+Currently, `NuGet.exe` [scans all directories in the `PATH` environment variable](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.CommandLine/MsBuildUtility.cs#L708-L736) to find `MSBuild.exe`.
+
+#### Plugin execution
+
+In the [current implementation](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Core/NuGet.Protocol/Plugins/PluginFactory.cs#L155-L181), NuGet launches the `.exe` file in a separate process when running on the .NET Framework.
+On the other hand, when running on .NET, NuGet executes the `dotnet plugin-in.dll` command in a separate process.
+
+The plan is to slightly adjust this implementation.
+NuGet will launch the `.exe` file in a separate process when running on Windows, irrespective of whether the runtime is .NET Framework or .NET.
+For non-Windows platforms, if the file does not have an extension, it will launch the executable directly in a separate process.
+If the file has a `.dll` extension, it will continue to execute the `dotnet plugin-name.dll` command in a separate process.
+It is worth noting that a plugin with a `.dll` extension will only be executed using the dotnet CLI if the file is located under the NuGet plugin-specific `netcore` subdirectory.
+This is to support backward compatibility with the NuGet plugins discovered from predefined plugin directories.
+As mentioned above, during PATH scanning, NuGet searches only for plugins with `.exe` or `.bat` extensions on Windows, and for files with the executable bit set on Linux/Mac.
+
+NuGet is designed to launch credential provider plugins sequentially, not simultaneously.
+This is to avoid multiple authentication prompts from different plugins at the same time.
+If a customer has installed both the legacy and the new .NET Tool credential providers, which can provide credentials for the same feed, NuGet follows a specific process.
+It launches one provider at a time, and if that provider successfully returns the credentials for the feed, NuGet will not need to invoke the other provider for the same feed.
+
+## Drawbacks
+
+- This approach doesn't support the installation of NuGet plugins as a [.NET local tool](https://learn.microsoft.com/en-us/dotnet/core/tools/local-tools-how-to-use).
+The reason for this is that running a local tool requires the invocation of the `dotnet tool run` command.
+However, in the current design, we have considered launching the .NET tool executable in a separate process without relying on the said command.
+See the `Future Possibilities` section for more details.
+
+- The discoverability of NuGet plugins published as .NET Tools is challenging for users because the `dotnet tool search` command only filters based on the `PackageType` being `DotnetTool`.
```

Comment:
I'm not concerned about this too much, since most of the time people will learn about the plugin from their feed provider.